File: 811 - Subdomain Visit Count.cpp
#include <vector>
#include <string>
#include <algorithm>
#include <utility>
#include <map>
using namespace std;


class Solution {
public:
    vector<string> subdomainVisits(vector<string>& cpdomains) {
        unordered_map<string, int> m;
        for (string & s : cpdomains) {
            int spaceidx = s.find(' ');
            int count = stoi(s.substr(0, spaceidx));
            string subdomains = s.substr(spaceidx + 1);
            m[subdomains] += count;
            int period = subdomains.size();
            string temp = "";
            do {
                period = subdomains.rfind('.', period - 1);
                if (period != string::npos) {
                    temp = subdomains.substr(period + 1);
                    m[temp] += count;
                }
            } while (period != string::npos);
        }
        vector<string> res;
        for (auto item : m) {
            string t = to_string(item.second) + " " + item.first;
            res.push_back(t);
        }
        return res;
    }
};

File: 811 - Subdomain Visit Count.js
/**
 * @param {string[]} cpdomains
 * @return {string[]}
 */
var subdomainVisits = function (cpdomains) {
    let m = new Map();
    for (let cpdomain of cpdomains) {
        let cnd = cpdomain.split(' ');
        let count = Number.parseInt(cnd[0]);
        if (!m.has(cnd[1])) {
            m.set(cnd[1], 0);
        }
        m.set(cnd[1], m.get(cnd[1]) + count);

        let subdomains = cnd[1].split('.');
        let temp = '';
        for (let i = subdomains.length - 1; i > 0; i--) {
            temp = i == subdomains.length - 1 ? subdomains[i] : subdomains[i] + '.' + temp;
            if (!m.has(temp)) {
                m.set(temp, 0);
            }
            m.set(temp, m.get(temp) + count);
        }
    }

    let res = []
    for (let dom of m.keys()) {
        let t = m.get(dom) + ' ' + dom;

        res.push(t);
    }
    return res;
};


let a = ["9001 discuss.leetcode.com"];
let r = subdomainVisits(a);
console.log(r);